-- build subject table
truncate table reporting.subject cascade
;
insert into reporting.subject
	(--subject_pk --autogenerated
	 subject_bk
	,disease_status)
select distinct
	 patient
    ,donor_disease_status
from staging.sample_tracker
;

-- sample table
truncate table reporting.sample cascade
;
insert into reporting.sample
  (--sample_pk --autogenerated
   subject_pk
  ,sample_bk
  ,amount
  ,unit
  ,collection
  ,organ
  ,organ_piece
  ,stage
  ,phenotype
  ,size_length
  ,size_width
  ,size_depth
  ,x_coord
  ,y_coord
  ,dysplasia_cat
  ,dysplasia_pct
  ,notes
  ,vap_names)
select
	 subject.subject_pk
	,st.sample_name
	,regexp_replace(case when st.amt_mg ~ '.+' then st.amt_mg when st.amt_ul ~ '.+' then st.amt_ul end, '[~>]', '')::numeric(10,2) as amt --get rid of approximate (~) and greater than (>) indicators in tracking sheet
    ,case when st.amt_mg ~ '.+' then 'mg' when st.amt_ul ~ '.+' then 'ul' end as unit
    ,lower(st.collection) as collection
    ,case
		when lower(st.sample_location) ~ 'descending|sigmoid|transverse|ascending|cecum|rectum'
        then 'colon'
        when lower(st.sample_location) ~ 'ileum|jejunum|mid-jejunum|duodenum'
        then 'small bowel'
	 end as organ
    ,case
    	when st.sample_location ~ '@'
    	then lower(substring(st.sample_location, 1, position('@' in st.sample_location)-1))
    	else lower(st.sample_location)
     end as organ_piece
    ,lower(st.stage) as stage
    ,lower(st.phenotype) as phenotype
    ,coalesce(nullif(split_part(st.size_mm, 'x', 1), ''), '1')::numeric(4,1) as size_length
    ,coalesce(nullif(split_part(st.size_mm, 'x', 2), ''), '1')::numeric(4,1) as size_width
    ,coalesce(nullif(split_part(st.size_mm, 'x', 3), ''), '1')::numeric(4,1) as size_depth
    ,coords.x
    ,coords.y
    ,lower(st.dysplasia_category) as dysplasia_category
    ,case
    	when st.dysplasia_percentage !~ '&'
    	then st.dysplasia_percentage::numeric(4,1)
	 end as dysplasia_percentage
    ,st.notes
    ,st.vap_names
from staging.sample_tracker st
join reporting.subject on subject.subject_bk = st.patient
left join staging.sample_coordinates coords on coords.sample_name = st.sample_name
where st.tissue_number like '%-1' --only get the first subsample (i.e. the original) for this table
;

-- tissue
truncate table reporting.tissue cascade
;
insert into reporting.tissue
  (--tissue_pk --autogenerated
   sample_pk
  ,tissue_bk
  ,init_pres_mthd
  ,curr_pres_mthd)
select
	 sample.sample_pk
	,st.tissue_number
	,lower(st.preservation_method)
    ,lower(st.current_state)
from staging.sample_tracker st
join reporting.sample on sample.sample_bk = st.sample_name
;

-- storage
truncate table reporting."storage" cascade
;
insert into reporting."storage"
	(--storage_pk --autogenerated
	 tissue_pk
	,building
	,freezer
	,rack
	,box_name)
select
	 tissue.tissue_pk
	,st.building
	,st.freezer
	,st.rack
	,st.box_name
from staging.sample_tracker st
join reporting.tissue on tissue.tissue_bk = st.tissue_number
;

-- assays
-- here we're manually unpivoting the individual assay columns into a "type" and
-- "details" pair before using regexes to pull out the fields we care about
drop table if exists assay_stg
;
create temp table assay_stg as
select
	 tissue_bk
    ,'DNA Whole Genome Sequencing' as "type"
    ,dna_wgs as details
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where dna_wgs ~ '.+'
union
select
	 tissue_bk
    ,'DNA Whole Exome Sequencing'
	,dna_wes
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where dna_wes ~ '.+'
union
select
	 tissue_bk
    ,'Whole Genome Bisulfite'
    ,wg_bisulfite
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where wg_bisulfite ~ '.+'
union
select
	 tissue_bk
    ,'RNA Sequencing'
	,rna_seq
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where rna_seq ~ '.+'
union
select
	 tissue_bk
    ,'Bulk ATAC'
    ,bulk_atac
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where bulk_atac ~ '.+'
union
select
	 tissue_bk
    ,'Single-Gland ATAC Whole Genome Sequencing'
    ,single_gland_atac_wgs
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where single_gland_atac_wgs ~ '.+'
union
select
	 tissue_bk
    ,'Single-Nucleus RNA Sequencing'
    ,sn_rna_seq
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where sn_rna_seq ~ '.+'
union
select
	 tissue_bk
    ,'Single-Nucleus ATAC Sequencing'
    ,sn_atac_seq
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where sn_atac_seq ~ '.+'
union
select
	 tissue_bk
    ,'Single-Cell ATACseq' as assay_type
    ,sc_atac_seq
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where sc_atac_seq ~ '.+'
union
select
	 tissue_bk
    ,'Single-Cell RNAseq' as assay_type
    ,sc_rna_seq
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where sc_rna_seq ~ '.+'
union
select
	 tissue_bk
    ,'HE/CODEX' as assay_type
	,he_codex
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where he_codex ~ '.+'
union
select
	 tissue_bk
    ,'Protein Isolation' as assay_type
	,protein_isolation
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where protein_isolation ~ '.+'
union
select
	 tissue_bk
    ,'RNA Isolation' as assay_type
    ,rna_isolation
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where rna_isolation ~ '.+'
union
select
	 tissue_bk
    ,'DNA Isolation' as assay_type
    ,dna_isolation
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where dna_isolation ~ '.+'
union
select
	 tissue_bk
    ,'Proteomics' as assay_type
    ,proteomics
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where proteomics ~ '.+'
union
select
	 tissue_bk
    ,'Lipidomics' as assay_type
    ,lipidomics
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where lipidomics ~ '.+'
union
select
	 tissue_bk
    ,'Metabolomics' as assay_type
    ,metabolomics
from staging.sample_tracker
join reporting.tissue on tissue.tissue_bk = sample_tracker.tissue_number
where metabolomics ~ '.+'
;

truncate table reporting.assay
;
insert into reporting.assay
	(--assay_pk --autogenerated
	 tissue_pk
	,assay_type
	,perf_date
	,researcher
	,equipment
	,seq_depth)
select
	 tissue.tissue_pk
	,assay_stg."type"
	,substring(lower(assay_stg.details), '(([1-9]|1[0-2])\/([1-9]|[1-2][0-9]|3[0-1])\/(1[5-9]|2[0-5]))')::date as perf_date
    ,substring(lower(assay_stg.details), 'tuhin|lihua|kevin|mark') as researcher
    ,substring(lower(assay_stg.details), 'novogene|personalis|bulk omniatac') as equipment
    ,substring(lower(assay_stg.details), '\d{2,3}x') as seq_depth
from assay_stg
join reporting.tissue on tissue.tissue_bk = assay_stg.tissue_bk
;
--pathology
truncate table reporting.pathology
;
insert into reporting.pathology
	(sample_pk
	,normal_tissue
	,polyp_type
	,architectural_distortion
	,villous_blunting
	,surface_and_crypt_epithelial_injury
	,lamina_propria_inflammation
	,intra_epithelial_infiltration
	,lymphocytic_infiltration
	,crypts
	,cryptitis
	,pseudomembrane
	,percent_of_acellular_mucin_by_area
	,dysplasia
	,degree_of_dysplasia
	,high
	,high_pct
	,low
	,low_pct
	,pct_neoplastic_nuclei_tumor_only
	,pct_non_neoplastic_stroma_by_area
	,pct_entire_tissue_involved_by_tumor
	,pct_normal_overall
	,pct_stroma_overall
	,pct_cancer_overall
	,pct_adenoma_overall
	,pct_stroma_in_cancer
	,pct_stroma_in_adenoma
	,pct_necrosis_in_tumor
	,cancer
	,cancer_type
	,grade
	,infiltrating_lymphocytes
	,crohn_like_response
	,necrosis
	,pct_tumor_necrosis_by_cellularity
	,lvsi
	,tumor_budding
	,tnm_score
	,stage
	,additional_details_comments)
select
	 sample.sample_pk
	,nullif(nullif(normal_tissue                      ,'na'),'n/a')::boolean
	,polyp_type
	,nullif(nullif(architectural_distortion           ,'na'),'n/a')::boolean
	,nullif(nullif(villous_blunting                   ,'na'),'n/a')::boolean
	,nullif(nullif(surface_and_crypt_epithelial_injury,'na'),'n/a')::boolean
	,nullif(nullif(lamina_propria_inflammation        ,'na'),'n/a')::boolean
	,nullif(nullif(intra_epithelial_infiltration      ,'na'),'n/a')::boolean
	,nullif(nullif(lymphocytic_infiltration           ,'na'),'n/a')::boolean
	,nullif(nullif(crypts                             ,'na'),'n/a')::boolean
	,nullif(nullif(cryptitis                          ,'na'),'n/a')::boolean
	,nullif(nullif(pseudomembrane                     ,'na'),'n/a')::boolean
	,nullif(nullif(percent_of_acellular_mucin_by_area ,'na'),'n/a')::decimal(3,2)
	,nullif(nullif(dysplasia                          ,'na'),'n/a')::boolean
	,degree_of_dysplasia
	,nullif(nullif(high                               ,'na'),'n/a')::boolean
	,(case
		when lower(high_and_low) like '%high%low%' then split_part(percentages_with_respect_to_previous_row, ',', 1)
		when lower(high_and_low) like '%low%high%' then split_part(percentages_with_respect_to_previous_row, ',', 2)
		when lower(high_and_low) like '%high%'     then percentages_with_respect_to_previous_row
	 end::integer/100.0)::decimal(3,2)
	,nullif(nullif(low                                ,'na'),'n/a')::boolean
	,(case
		when lower(high_and_low) like '%high%low%' then split_part(percentages_with_respect_to_previous_row, ',', 2)
		when lower(high_and_low) like '%low%high%' then split_part(percentages_with_respect_to_previous_row, ',', 1)
		when lower(high_and_low) like '%low%'      then percentages_with_respect_to_previous_row
	 end::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_of_neoplastic_cell_nuclei_as_a_total_of_all_cell_nuclei,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_of_non_neoplastic_stroma_by_area,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_of_entire_tissue_involved_by_tumor,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_normal_overall,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_stroma_overall,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_cancer_overall,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_adenoma_overall,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_stroma_in_cancer,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_stroma_in_adenoma,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_necrosis_in_tumor,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,nullif(nullif(cancer,'na'),'n/a')::boolean
	,nullif(nullif(cancer_type,'na'),'n/a')
	,nullif(nullif(grade,'na'),'n/a')::smallint
	,nullif(nullif(infiltrating_lymphocytes,'na'),'n/a')::boolean
	,nullif(nullif(crohn_like_response     ,'na'),'n/a')::boolean
	,nullif(nullif(necrosis                ,'na'),'n/a')::boolean
	,(nullif(nullif(percent_of_tumor_necrosis_by_cellularity,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,nullif(nullif(lvsi                    ,'na'),'n/a')::boolean
	,nullif(nullif(tumor_budding              ,'na'),'n/a')
	,nullif(nullif(tnm_score                  ,'na'),'n/a')
	,nullif(nullif(pathology.stage            ,'na'),'n/a')
	,nullif(nullif(additional_details_comments,'na'),'n/a')
from staging.pathology
join reporting.sample on sample.sample_bk = pathology.sample
;
--ATAC-seq bulk metadata
truncate table reporting.atacseq_bulk_metadata
;
insert into reporting.atacseq_bulk_metadata
  (sample_bk
  ,hubmap_subject_bk
  ,hubmap_tissue_bk
  ,execution_datetime
  ,protocols_io_doi
  ,operator
  ,operator_email
  ,pi
  ,pi_email
  ,assay_category
  ,assay_type
  ,analyte_class
  ,is_targeted
  ,acquisition_instrument_vendor
  ,acquisition_instrument_model
  ,is_technical_replicate
  ,library_id
  ,bulk_atac_cell_isolation_protocols_io_doi
  ,nuclei_quality_metric
  ,bulk_transposition_input_number_nuclei
  ,transposition_method
  ,transposition_transposase_source
  ,transposition_kit_number
  ,library_construction_protocols_io_doi
  ,library_layout
  ,library_adapter_sequence
  ,library_pcr_cycles
  ,library_average_fragment_size
  ,library_creation_date
  ,sequencing_reagent_kit
  ,sequencing_read_format
  ,sequencing_read_percent_q30
  ,sequencing_phix_percent
  ,library_final_yield_value
  ,library_final_yield_units
  ,library_preparation_kit
  ,library_concentration_value
  ,library_concentration_unit
  ,metadata_path
  ,data_path
  ,fastq_file_size)
select
   --sample.sample_pk
  sample_bk
  ,hubmap_subject_bk
  ,hubmap_tissue_bk
  ,execution_datetime::timestamp
  ,protocols_io_doi
  ,operator
  ,operator_email
  ,pi
  ,pi_email
  ,assay_category
  ,assay_type
  ,analyte_class
  ,is_targeted::boolean
  ,acquisition_instrument_vendor
  ,acquisition_instrument_model
  ,is_technical_replicate::boolean
  ,library_id
  ,bulk_atac_cell_isolation_protocols_io_doi
  ,nuclei_quality_metric
  ,bulk_transposition_input_number_nuclei::integer
  ,transposition_method
  ,transposition_transposase_source
  ,transposition_kit_number::integer
  ,library_construction_protocols_io_doi
  ,library_layout
  ,library_adapter_sequence
  ,library_pcr_cycles::integer
  ,library_average_fragment_size::numeric(7,3)
  ,library_creation_date::date
  ,sequencing_reagent_kit
  ,sequencing_read_format
  ,sequencing_read_percent_q30::numeric(7,3)
  ,sequencing_phix_percent::numeric(7,3)
  ,library_final_yield_value::numeric(7,3)
  ,library_final_yield_units
  ,library_preparation_kit
  ,library_concentration_value::numeric(7,3)
  ,library_concentration_unit
  ,metadata_path
  ,data_path
  ,fastq_file_size
from staging.atacseq_bulk_metadata
join reporting.sample on sample.sample_bk = atacseq_bulk_metadata.sample_bk
