-- study --
-- 6/23/20 (DLC) - This is hardcoded for now just for simplicity, but in the
--                 future it will be necessary to decide dynamically based
--                 on which document is being imported or other factors.  
insert into core.study
  (--study_pk --autogenerated
   study_bk)
select 'HuBMAP'
;

-- subject --
insert into core.subject
  (--subject_pk --autogenerated
   study_pk
  ,subject_bk
  ,disease_status
  ,rectum_position
  ,descending_position
  ,transverse_position
  ,ascending_position
  ,colon_length)
select distinct
   study.study_pk
  ,patient
  ,donor_disease_status
  ,rectum_position::numeric(4,1)
  ,descending_position::numeric(4,1)
  ,transverse_position::numeric(4,1)
  ,ascending_position::numeric(4,1)
  ,colon_length::numeric(4,1)
from staging.sample_tracker
join core.study on study.study_bk = 'HuBMAP' --see note above about hardcoding
join staging.colon_measurements cm on cm.subject_bk = patient
;

-- sample --
insert into core.sample
  (--sample_pk --autogenerated
   subject_pk
  ,sample_bk
  ,amount
  ,unit
  ,collection
  ,organ
  ,organ_piece
  ,stage
  ,phenotype
  ,size_length
  ,size_width
  ,size_depth
  ,x_coord
  ,y_coord
  ,dysplasia_cat
  ,dysplasia_pct
  ,notes
  ,vap_names
  ,atacseq_bulk
  ,atacseq_sn
  ,atacseq_sc
  ,codex
  ,lipidomics
  ,metabolomics
  ,proteomics
  ,rnaseq_bulk
  ,rnaseq_sn
  ,rnaseq_sc
  ,isolation_dna
  ,isolation_rna
  ,isolation_protein
  ,single_gland_atac_wgs
  ,wes
  ,wgs
  ,wg_bisulfite)
select
	 subject.subject_pk
	,st.sample_name
	,regexp_replace(case when st.amt_mg ~ '.+' then st.amt_mg when st.amt_ul ~ '.+' then st.amt_ul end, '[~>]', '')::numeric(10,2) as amt --get rid of approximate (~) and greater than (>) indicators in tracking sheet
    ,case when st.amt_mg ~ '.+' then 'mg' when st.amt_ul ~ '.+' then 'ul' end as unit
    ,lower(st.collection) as collection
    ,case
		when lower(st.sample_location) ~ 'descending|sigmoid|transverse|ascending|cecum|rectum'
        then 'colon'
        when lower(st.sample_location) ~ 'ileum|jejunum|mid-jejunum|duodenum'
        then 'small bowel'
	 end as organ
    ,case
    	when st.sample_location ~ '@'
    	then lower(substring(st.sample_location, 1, position('@' in st.sample_location)-1))
    	else lower(st.sample_location)
     end as organ_piece
    ,lower(st.stage) as stage
    ,lower(st.phenotype) as phenotype
    ,nullif(split_part(st.size_mm, 'x', 1), '')::numeric(4,1) as size_length
    ,nullif(split_part(st.size_mm, 'x', 2), '')::numeric(4,1) as size_width
    ,nullif(split_part(st.size_mm, 'x', 3), '')::numeric(4,1) as size_depth
    ,coords.x::numeric(5,1)
    ,coords.y::numeric(5,1)
    ,lower(st.dysplasia_category) as dysplasia_category
    ,case
    	when st.dysplasia_percentage !~ '&'
    	then st.dysplasia_percentage::numeric(4,1)
	 end as dysplasia_percentage
    ,st.notes
    ,st.vap_names
    --boolean-valued columns for whether an assay was performed
    ,(case when nullif(st.bulk_atac,'')             is not null then 1 else 0 end)::bool
    ,(case when nullif(st.sn_atac_seq,'')           is not null then 1 else 0 end)::bool
    ,(case when nullif(st.sc_atac_seq,'')           is not null then 1 else 0 end)::bool
    ,(case when nullif(st.he_codex,'')              is not null then 1 else 0 end)::bool
    ,(case when nullif(st.lipidomics,'')            is not null then 1 else 0 end)::bool
    ,(case when nullif(st.metabolomics,'')          is not null then 1 else 0 end)::bool
    ,(case when nullif(st.proteomics,'')            is not null then 1 else 0 end)::bool
    ,(case when nullif(st.rna_seq,'')               is not null then 1 else 0 end)::bool
    ,(case when nullif(st.sn_rna_seq,'')            is not null then 1 else 0 end)::bool
    ,(case when nullif(st.sc_rna_seq,'')            is not null then 1 else 0 end)::bool
    ,(case when nullif(st.dna_isolation,'')         is not null then 1 else 0 end)::bool
    ,(case when nullif(st.rna_isolation,'')         is not null then 1 else 0 end)::bool
    ,(case when nullif(st.protein_isolation,'')     is not null then 1 else 0 end)::bool
    ,(case when nullif(st.single_gland_atac_wgs,'') is not null then 1 else 0 end)::bool
    ,(case when nullif(st.dna_wgs,'')               is not null then 1 else 0 end)::bool
    ,(case when nullif(st.dna_wes,'')               is not null then 1 else 0 end)::bool
    ,(case when nullif(st.wg_bisulfite,'')          is not null then 1 else 0 end)::bool
from staging.sample_tracker st
join core.subject on subject.subject_bk = st.patient
left join staging.sample_coordinates coords on coords.sample_name = st.sample_name
where st.tissue_number like '%-1' --only get the first subsample (i.e. the original) for this table
;

-- tissue --
insert into core.tissue
  (--tissue_pk --autogenerated
   sample_pk
  ,tissue_bk
  ,init_pres_mthd
  ,curr_pres_mthd)
select
	 sample.sample_pk
	,st.tissue_number
	,lower(st.preservation_method)
    ,lower(st.current_state)
from staging.sample_tracker st
join core.sample on sample.sample_bk = st.sample_name
;

-- storage --
insert into core."storage"
	(--storage_pk --autogenerated
	 tissue_pk
	,building
	,freezer
	,rack
	,box_name)
select
	 tissue.tissue_pk
	,st.building
	,st.freezer
	,st.rack
	,st.box_name
from staging.sample_tracker st
join core.tissue on tissue.tissue_bk = st.tissue_number
;

-- assay --
insert into core.assay
	(--assay_pk --autogenerated
	 tissue_pk
	,assay_type
	,experiment_date
	,researcher)
select tissue.tissue_pk
	,l.k as assay_type
	,case
		when l.v ~ '(([1-9]|1[0-2])\/([1-9]|[1-2][0-9]|3[0-1])\/(1[5-9]|2[0-5]))'
		then substring(lower(l.v), '(([1-9]|1[0-2])\/([1-9]|[1-2][0-9]|3[0-1])\/(1[5-9]|2[0-5]))')::date
	 end as experiment_date
	,case
		when lower(l.v) ~ 'tuhin|lihua|kevin|mark' 
		then substring(lower(l.v), 'tuhin|lihua|kevin|mark') 
	 end as researcher
from staging.sample_tracker
join core.tissue on tissue.tissue_bk = sample_tracker.tissue_number
join lateral(values
	 ('DNA Whole Genome Sequencing', sample_tracker.dna_wgs)
	,('DNA Whole Exome Sequencing', sample_tracker.dna_wes)
	,('Whole Genome Bisulfite', sample_tracker.wg_bisulfite)
	,('RNA Sequencing', sample_tracker.rna_seq)
	,('Bulk ATAC', sample_tracker.bulk_atac)
	,('Single-Gland ATAC Whole Genome Sequencing', sample_tracker.single_gland_atac_wgs)
	,('Single-Nucleus RNA Sequencing', sample_tracker.sn_rna_seq)
	,('Single-Nucleus ATAC Sequencing', sample_tracker.sn_atac_seq)
	,('Single-Cell ATACseq', sample_tracker.sc_atac_seq)
	,('Single-Cell RNAseq', sample_tracker.sc_rna_seq)
	,('HE/CODEX', sample_tracker.he_codex)
	,('Protein Isolation', sample_tracker.protein_isolation)
	,('RNA Isolation', sample_tracker.rna_isolation)
	,('DNA Isolation', sample_tracker.dna_isolation)
	,('Proteomics', sample_tracker.proteomics)
	,('Lipidomics', sample_tracker.lipidomics)
	,('Metabolomics', sample_tracker.metabolomics)
) l(k,v) on l.v is not null
;

-- pathology --
insert into core.pathology
	(sample_pk
	,normal_tissue
	,polyp_type
	,architectural_distortion
	,villous_blunting
	,surface_and_crypt_epithelial_injury
	,lamina_propria_inflammation
	,intra_epithelial_infiltration
	,lymphocytic_infiltration
	,crypts
	,cryptitis
	,pseudomembrane
	,percent_of_acellular_mucin_by_area
	,dysplasia
	,degree_of_dysplasia
	,high
	,high_pct
	,low
	,low_pct
	,pct_neoplastic_nuclei_tumor_only
	,pct_non_neoplastic_stroma_by_area
	,pct_entire_tissue_involved_by_tumor
	,pct_normal_overall
	,pct_stroma_overall
	,pct_cancer_overall
	,pct_adenoma_overall
	,pct_stroma_in_cancer
	,pct_stroma_in_adenoma
	,pct_necrosis_in_tumor
	,cancer
	,cancer_type
	,grade
	,infiltrating_lymphocytes
	,crohn_like_response
	,necrosis
	,pct_tumor_necrosis_by_cellularity
	,lvsi
	,tumor_budding
	,tnm_score
	,stage
	,additional_details_comments)
select
	 sample.sample_pk
	,nullif(nullif(normal_tissue                      ,'na'),'n/a')::boolean
	,polyp_type
	,nullif(nullif(architectural_distortion           ,'na'),'n/a')::boolean
	,nullif(nullif(villous_blunting                   ,'na'),'n/a')::boolean
	,nullif(nullif(surface_and_crypt_epithelial_injury,'na'),'n/a')::boolean
	,nullif(nullif(lamina_propria_inflammation        ,'na'),'n/a')::boolean
	,nullif(nullif(intra_epithelial_infiltration      ,'na'),'n/a')::boolean
	,nullif(nullif(lymphocytic_infiltration           ,'na'),'n/a')::boolean
	,nullif(nullif(crypts                             ,'na'),'n/a')::boolean
	,nullif(nullif(cryptitis                          ,'na'),'n/a')::boolean
	,nullif(nullif(pseudomembrane                     ,'na'),'n/a')::boolean
	,nullif(nullif(percent_of_acellular_mucin_by_area ,'na'),'n/a')::decimal(3,2)
	,nullif(nullif(dysplasia                          ,'na'),'n/a')::boolean
	,degree_of_dysplasia
	,nullif(nullif(high                               ,'na'),'n/a')::boolean
	,(case
		when lower(high_and_low) like '%high%low%' then split_part(percentages_with_respect_to_previous_row, ',', 1)
		when lower(high_and_low) like '%low%high%' then split_part(percentages_with_respect_to_previous_row, ',', 2)
		when lower(high_and_low) like '%high%'     then percentages_with_respect_to_previous_row
	 end::integer/100.0)::decimal(3,2)
	,nullif(nullif(low                                ,'na'),'n/a')::boolean
	,(case
		when lower(high_and_low) like '%high%low%' then split_part(percentages_with_respect_to_previous_row, ',', 2)
		when lower(high_and_low) like '%low%high%' then split_part(percentages_with_respect_to_previous_row, ',', 1)
		when lower(high_and_low) like '%low%'      then percentages_with_respect_to_previous_row
	 end::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_of_neoplastic_cell_nuclei_as_a_total_of_all_cell_nuclei,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_of_non_neoplastic_stroma_by_area,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_of_entire_tissue_involved_by_tumor,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_normal_overall,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_stroma_overall,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_cancer_overall,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_adenoma_overall,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_stroma_in_cancer,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_stroma_in_adenoma,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,(nullif(nullif(percent_necrosis_in_tumor,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,nullif(nullif(cancer,'na'),'n/a')::boolean
	,nullif(nullif(cancer_type,'na'),'n/a')
	,nullif(nullif(grade,'na'),'n/a')::smallint
	,nullif(nullif(infiltrating_lymphocytes,'na'),'n/a')::boolean
	,nullif(nullif(crohn_like_response     ,'na'),'n/a')::boolean
	,nullif(nullif(necrosis                ,'na'),'n/a')::boolean
	,(nullif(nullif(percent_of_tumor_necrosis_by_cellularity,'na'),'n/a')::integer/100.0)::decimal(3,2)
	,nullif(nullif(lvsi                    ,'na'),'n/a')::boolean
	,nullif(nullif(tumor_budding              ,'na'),'n/a')
	,nullif(nullif(tnm_score                  ,'na'),'n/a')
	,nullif(nullif(pathology.stage            ,'na'),'n/a')
	,nullif(nullif(additional_details_comments,'na'),'n/a')
from staging.pathology
join core.sample on sample.sample_bk = pathology.sample
;
